docs/mskimmer.1
.TH mskimmer 1 "February 2003" Linux "Linux Audio Manual"
.SH NAME
mskimmer \- The Call Commander Call Skimmer

.SH SYNOPSIS
mskimmer [OPTION] [config-file]

.SH DESCRIPTION
\fBmskimmer\fP is a specialized audio recorder that uses one or more audio
adapters to record telephone conversations.  Recordings are done under
the control of an associated telephone interface, as configured by
\fBmskimmer\fP's user interface and configuration file.

\fBMskimmer\fP features two recording decks, one for each telephone bank.
Each deck consists of a set of line buttons, a RECORD button, a STOP
button, a record timer and a file name display.  Selecting a line
button has the effect of causing the associated deck to begin
recording should that line be placed on the air on the appropriate
bank.  When all selected lines are taken off air (either by being
placed on hold, or by being hung up), the recording stops.  It is also
possible to operate the deck manually by means of the associated
RECORD and STOP buttons.

.SH OPTIONS
.TP
.B -v, --version
output version information and exit.

.SH CONFIGURATION FILE 
Upon startup, \fBmskimmer\fP will look for a configuration file in the
following locations/order.
.RS
.TP
Command Line
The file whose name is passed as the last argument on the command
line.
.TP
$HOME/.mlrc
The file \fB.mlrc\fP in the user's home directory.
.TP
/etc/ml.conf
The file \fBml.conf\fP in the /etc/ directory.
.RE
.RE

If no configuration file is found at any of the above locations,
\fBmskimmer\fP will use a default configuration compiled into the program.
When a configuration file is found, further searching
stops.  The format of this file is documented in the ml.conf(5) man
page.  This file is also used by the Call Commander user interface
program, \fBml(1)\fP.  The \fI[Skimmer]\fP section of the
configuration file is used to control \fBmskimmer(1)\fP's operation.

.SH BUGS
On the Telos TWOx12, password support is not implemented as of firmware
version 2.02.  Thus, attempts to connect to this unit should \fIalways\fP
use an empty password.

.SH AUTHOR
Fred Gleason <fredg@paravelsystems.com>
.SH "SEE ALSO"
.BR callcommander(1),
.BR ml(1),
.BR ml.conf(5),
.BR mld(8),
.BR mld.conf(5)










NEWS
The NEWS file for the Call Commander package.

-------------------------------------------------------------------------------
v2.1.1 -- 04/26/2013

Changes in this version include:

      Message Box Right-Click Menu.  Added a 'Clear' item to the menu
      that clears the contents of the Message Box (both local and
      remote).

      Graphical Cleanups.  Fixed multiple issues with rendering buttons
      and other graphical elements under various environments (OS X,
      GNOME).

      Various bugfixes.  The the ChangeLog for specifics.
-------------------------------------------------------------------------------
v2.1.0 -- 04/15/2013

Changes in this version include:

      Modified password storage in the database to use cleartext entries.
      Be sure to clear the password on at least one account with
      Admin rights before applying the update, then use this account
      to reset all passwords after the update.
-------------------------------------------------------------------------------
v2.0.6 -- 08/24/2012

Changes in this version include:

      System-Wide Connection Manager Entries (Windows only).  It is now
      possible to configure CallCommander to store and retrieve Connection
      Manager entries to/from a common system-wide directory, rather than
      on a per-user basis by changing the value of the registry key
      \\HKEY_LOCAL_MACHINE\SOFTWARE\SalemRadioLabs\CallCommander\
      ConnectionManagerScope from 'User' to 'System'.
-------------------------------------------------------------------------------
v2.0.1 -- 08/18/2011

Changes in this version include:

      Qt port from 3.x to 4.x.

      Mac OS X Port.
-------------------------------------------------------------------------------
v1.5.31 -- 09/08/2009

Changes in this version include:

      Delay Button Display Options.  It is now possible to configure the delay
      button in ml(1) to display the current delay offset value by use of the
      'ShowDelay=' option in the [DelayButton] section of ml.conf(5).  The
      delay button can now be made to flash when the delay offset falls below
      the value set in 'DelayThreshold='.

      Message Box Changes.  It is now possible to paste large text messages
      through the message box without having the message split into smaller
      sections.
-------------------------------------------------------------------------------
v1.5.18 -- 05/28/2008

Changes in this version include:

      Asterisk PBX Support (EXPERIMENTAL).  It is now possible to use a
      suitably configured Asterisk system as a broadcast telephone interface,
      with CallCommander as the control interface.  Further information can be
      found in the 'ASTERISK.txt' file.

      Ml.conf(5) configuration directives.  A new 'SendTimeOffset=' parameter
      has been added to ml.conf(5) to allow cusotmization of the relaying of
      time offsets between ml(1) instances.  See the ml.conf(5) man page for
      details.
-------------------------------------------------------------------------------
v1.5.3 -- 05/30/2006

This is the fourth beta release of the "new" CallCommander.
Additional changes over the previous v1.5.2 release include:

      Online Users List.  A new button, titled 'Who's Here?', has been
      added to the default CallCommander desktop that brings up a
      modeless dialog showing information regarding all users
      currently connected.  This button can be configured for custom
      layouts like any other control -- see the [ClientButton] section
      of the ml.conf(5) man page for details.

      CallerID Monitoring Tweaks.  The operation of mcidmd(8) has been
      modified so that connections to the host phone system are made
      only when a given show is active, so as to avoid exhausting the
      pool of available connections for certain phone systems
      (e.g. the Telos TWOx12).

      Caller Metadata Refreshing.  Late-joining ml(1) clients are now
      automatically refreshed to receive any caller data already
      entered from other clients.

      Various Bugfixes.  See the ChangeLog for details.

IMPORTANT NOTE:  The v1.5.x series of CallCommander versions are *not*
backward compatible with the previous 1.3.x and 1.1.x versions.  When
upgrading from a 1.3.x or 1.1.x version, all server and client
components should be upgraded simultaneously.

Library Versions:
   This version requires that, at a minimum, libradio-0.97.0,
   librhpi-0.94.6 and libtelos2101-1.2.0 be installed.  If installing
   from RPM, the version of the currently installed libraries can be 
   determined by doing:

      rpm -q libradio
      rpm -q libtelos2101
-------------------------------------------------------------------------------
v1.5.2 --05/19/2006

This is the third beta release of the "new" CallCommander.
Additional changes over the previous v1.5.1 release include:

      Ease Of Use Improvements.  The Connection Manager has been
      tweaked to make building connections to certain phone system
      types (e.g. Telos TWOx12) easier and more intuitive.

      Robustness Tweaks.  The two server components [mld(8) and
      mcidmd(8)] have been overhauled with a view to making their
      operation reliable even in the face of intermittent/rebooted
      phone system hardware.

      Miscellaneous Bugfixes.  See the ChangeLog for details.

IMPORTANT NOTE:  The v1.5.x series of CallCommander versions are *not*
backward compatible with the previous 1.3.x and 1.1.x versions.  When
upgrading from a 1.3.x or 1.1.x version, all server and client
components should be upgraded simultaneously.

Library Versions:
   This version requires that, at a minimum, libradio-0.97.0,
   librhpi-0.94.6 and libtelos2101-1.1.0 be installed.  If installing
   from RPM, the version of the currently installed libraries can be 
   determined by doing:

      rpm -q libradio
      rpm -q libtelos2101
-------------------------------------------------------------------------------

mcs/mskimmer.cpp
// mcs.cpp
//
// A Qt-based application for recording telephone calls
//
//   (C) Copyright 2002-2003 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <stdlib.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <QtGui/QApplication>
#include <QtCore/QSignalMapper>
#include <QtGui/QMessageBox>
#include <QtGui/QLabel>

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <globals.h>
#include <askbox.h>
#include <mskimmer.h>
#include <list_connections.h>

//
// Global Resources
//
MlConfig *ml_config;

//
// Icons
//
#include "../icons/callcommander-22x22.xpm"


void SigHandler(int signo)
{
  pid_t pLocalPid;

  switch(signo) {
  case SIGCHLD:
    pLocalPid=waitpid(-1,NULL,WNOHANG);
    while(pLocalPid>0) {
      pLocalPid=waitpid(-1,NULL,WNOHANG);
    }
    signal(SIGCHLD,SigHandler);
    return;
  }
}


MainWidget::MainWidget(QWidget *parent,const char *name)
  :QWidget(parent,name)
{
  struct stat conf_stat;
  bool file_found=false;
  char temp[256];
  bool ask_for_profile=true;
  QString filename;

  setCaption("Call Skimmer");

  mcs_connection=new ConnectionData(ConnectionData::StorageFilesystem);

  //
  // Determine Which Configuration File to Use
  //
  // First, the command line argument
  //
  if(qApp->argc()>1) {
    memset(&conf_stat,0,sizeof(struct stat));
    if(stat(qApp->argv()[qApp->argc()-1],&conf_stat)==0) {
      mcs_config=qApp->argv()[qApp->argc()-1];
      file_found=true;
    }
  }

  //
  // Next, the .mlrc File
  //
  if(!file_found) {
    if(strlen(getenv("HOME"))<259) {
      sprintf(temp,"%s/.mlrc",getenv("HOME"));
      if(stat(temp,&conf_stat)==0) {
	mcs_config=temp;
	file_found=true;
      }
    }
  }

  //
  // Finally, the global configuration
  //
  if(!file_found) {
    mcs_config=DEFAULT_CONFIG_FILE;
  }

  //
  // Create And Set Icon
  //
  mcs_callcommander_map=new QPixmap(callcommander_xpm);
  setIcon(*mcs_callcommander_map);

  //
  // Global Section
  //
  ml_config=new MlConfig(mcs_config);
  ml_config->load();

  //
  // Try to load the specified connection profile
  //
  if(!ml_config->connectionName().isEmpty()) {
    mcs_connection->setName(ml_config->connectionName());
    ask_for_profile=!mcs_connection->load();
  }

  //
  // Connection Dialog
  //
  if(ask_for_profile) {
    ListConnections *list=
      new ListConnections(mcs_connection,tr("Open\nConnection"),
			  ml_config,this);
    if(list->exec()<0) {
      exit(0);
    }
  }

  //
  // The Bus Driver
  //
  switch(mcs_connection->systemType()) {
      case ConnectionData::TypeAsterisk:
	mcs_driver=new AsteriskDriver(0,this,"mcs_driver");
	break;

      case ConnectionData::TypeTelos2101:
	mcs_driver=new Telos2101Driver(0,this,"mcs_driver");
	break;

      case ConnectionData::TypeTelos100Key:
	mcs_driver=
	  new Telos100Driver(Telos100Driver::KeyType,0,this,"mcs_driver");
	break;

      case ConnectionData::TypeTelos100Dim:
	mcs_driver=
	  new Telos100Driver(Telos100Driver::DimType,0,this,"mcs_driver");
	break;

      case ConnectionData::TypeTelos100OneXSix:
	mcs_driver=
	  new Telos100Driver(Telos100Driver::OneXSixType,0,this,
			     "mcs_driver");
	break;

      case ConnectionData::TypeGentnerTs612:
	mcs_driver=new Ts612Driver(0,this,"mcs_driver");
	break;

      case ConnectionData::TypeVirtual:
	mcs_driver=new VirtualDriver(0,this,"mcs_driver");
	break;

      case ConnectionData::TypeUnknown:
	QMessageBox::warning(this,tr("System Error"),
			     tr("Unknown System Type!"));
	exit(256);
  }
  mcs_driver->setClientType(ClientData::TypeAudioRecorder);
  connect(mcs_driver,SIGNAL(connected(unsigned,bool)),
	  this,SLOT(connectedData(unsigned,bool)));

  switch(mcs_connection->systemType()) {
      case ConnectionData::TypeVirtual:
      case ConnectionData::TypeTelos2101:
      case ConnectionData::TypeAsterisk:
	mcs_driver->connect(mcs_connection->hostName(),
			    mcs_connection->userName(),
			    mcs_connection->userPassword(),
			    mcs_connection->tcpPort(),
			    mcs_connection->showName(),
			    mcs_connection->showPassword(),
			    mcs_connection->location(),
			    mcs_connection->console(),
			    mcs_connection->consoleName());
	break;

      case ConnectionData::TypeTelos100Key:
      case ConnectionData::TypeTelos100Dim:
      case ConnectionData::TypeTelos100OneXSix:
      case ConnectionData::TypeGentnerTs612:
	mcs_driver->connect(mcs_connection->ttyPort(),
			    mcs_connection->showName(),
			    mcs_connection->showPassword(),
			    mcs_connection->location(),
			    mcs_connection->console(),
			    mcs_connection->consoleName());
	break;

      case ConnectionData::TypeUnknown:
	QMessageBox::warning(this,tr("System Error"),
			     tr("Unknown System Type!"));
	exit(256);
  }

  //
  // Setup Signal Handling
  //
  signal(SIGCHLD,SigHandler);

  //
  // Generate Title Font
  //
  QFont font("Helvetica",18,QFont::Bold);
  font.setPixelSize(18);

  //
  // Record Widgets
  //
  QLabel *label=new QLabel("Main Bank",this,"main_bank_label");
  label->setGeometry(10,4,sizeHint().width()-20,20);
  label->setFont(font);
  label->show();
  
  main_widget=new McsWidget(0,mcs_connection,mcs_driver,
			    this,"main_widget");
  main_widget->setGeometry(0,20,main_widget->sizeHint().width(),
			   main_widget->sizeHint().height());
  
  label=new QLabel("Conference Bank",this,"conf_bank_label");
  label->setGeometry(10,main_widget->sizeHint().height()+30,
		     sizeHint().width()-20,20);
  label->setFont(font);
  label->show();
  
  conf_widget=new McsWidget(1,mcs_connection,mcs_driver,
			    this,"conf_widget");
  conf_widget->setGeometry(0,50+main_widget->sizeHint().height(),
			   conf_widget->sizeHint().width(),
			   conf_widget->sizeHint().height());
  main_widget->show();
  conf_widget->show();
  
  //
  // Generate Button Font
  //
  font=QFont("Helvetica",12,QFont::Bold);
  font.setPixelSize(12);
  
  //
  // The Close Button
  //
  QPushButton *close_button=new QPushButton(this,"close_button");
  close_button->setGeometry(200,main_widget->sizeHint().height()+
			    conf_widget->sizeHint().height()+55,70,40);
  close_button->setFont(font);
  close_button->setText("&Close");
  connect(close_button,SIGNAL(clicked()),this,SLOT(closeData()));
  close_button->show();

  mcs_driver->requestLineState();
}


QSize MainWidget::sizeHint() const
{
  return QSize(285,450);
  return QSize(main_widget->sizeHint().width(),
	       main_widget->sizeHint().height()+
	       conf_widget->sizeHint().height()+100);
}


QSizePolicy MainWidget::sizePolicy() const
{
  return QSizePolicy(QSizePolicy::Fixed,QSizePolicy::Fixed);
}


void MainWidget::connectedData(unsigned id,bool state)
{
  if(state) {
    mcs_driver->connectToShow(mcs_connection->showName(),
			      mcs_connection->showPassword(),
			      mcs_connection->location());
  }
  else {
    QMessageBox::warning(this,"Login Failed","Login failed!");
    qApp->quit();
  }
}


void MainWidget::closeData()
{
  qApp->quit();
}


int main(int argc,char *argv[])
{
  QApplication a(argc,argv);
  
#ifndef WIN32
  for(int i=0;i<argc;i++) {
    if((!strcmp(argv[i],"-v"))||(!strcmp(argv[i],"--version"))) {
      printf("CallCommander Call Skimmer - version %s\n",VERSION);
      exit(0);
    }
  }
#endif  // WIN32

  MainWidget *w=new MainWidget(NULL,"main");
  a.setMainWidget(w);
  w->show();
  return a.exec();
}



